File: clients/go/error_guides_test.go
package fibril

import (
	"context"
	"encoding/json"
	"net"
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"
)

// The shared clients/error_guides.json is the single source of truth for the
// wording every client must carry for the errors it raises itself, the
// error-message analogue of wire_vectors.json.
// mustContain returns the required substrings for a guide case. The file also
// holds a string-valued "_note" key, so entries are parsed individually.
func mustContain(t *testing.T, name string) []string {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join("..", "error_guides.json"))
	if err != nil {
		t.Fatalf("read error_guides.json: %v", err)
	}
	var m map[string]json.RawMessage
	if err := json.Unmarshal(raw, &m); err != nil {
		t.Fatalf("parse error_guides.json: %v", err)
	}
	var entry struct {
		MustContain []string `json:"must_contain"`
	}
	if err := json.Unmarshal(m[name], &entry); err != nil {
		t.Fatalf("parse guide %q: %v", name, err)
	}
	return entry.MustContain
}

func assertContainsAll(t *testing.T, name, message string, want []string) {
	t.Helper()
	lower := strings.ToLower(message)
	for _, sub := range want {
		if !strings.Contains(lower, strings.ToLower(sub)) {
			t.Errorf("%s message missing %q\n  got: %s", name, sub, message)
		}
	}
}

func TestConnectionRefusedCarriesGuide(t *testing.T) {
	// An address with nothing listening: bind an ephemeral port, then free it.
	l, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatalf("listen: %v", err)
	}
	addr := l.Addr().String()
	_ = l.Close()

	_, err = Connect(context.Background(), addr, EngineOptions{ClientName: "go-test", HeartbeatInterval: time.Hour})
	if err == nil {
		t.Fatal("expected a connection-refused error")
	}
	assertContainsAll(t, "connection_refused", err.Error(), mustContain(t, "connection_refused"))
}

func TestHeartbeatTimeoutCarriesGuide(t *testing.T) {
	assertContainsAll(t, "heartbeat_timeout", heartbeatTimeoutMessage, mustContain(t, "heartbeat_timeout"))
}

File: clients/typescript/tests/error-guides.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { createServer } from "node:net";
import { readFileSync } from "node:fs";
import { ClientOptions } from "../src/index.js";
import { deserializeByContentType } from "../src/message.js";

// Shared source of truth for client-local guide wording (see
// clients/error_guides.json), the error-message analogue of wire_vectors.json.
const guides = JSON.parse(
  readFileSync(new URL("../../error_guides.json", import.meta.url), "utf8"),
) as Record<string, { must_contain: string[] }>;

async function closedPort(): Promise<number> {
  return new Promise((resolve) => {
    const srv = createServer();
    srv.listen(0, "127.0.0.1", () => {
      const port = (srv.address() as { port: number }).port;
      srv.close(() => resolve(port));
    });
  });
}

test("connection_refused carries the shared guide", async () => {
  const port = await closedPort();
  let message = "";
  try {
    await new ClientOptions().disableAutoReconnect().connect({ host: "127.0.0.1", port });
    assert.fail("connecting to a closed port must throw");
  } catch (err) {
    message = String((err as Error).message).toLowerCase();
  }
  for (const keyword of guides.connection_refused.must_contain) {
    assert.ok(
      message.includes(keyword.toLowerCase()),
      `connection_refused guide is missing ${keyword}: ${message}`,
    );
  }
});

test("decode_malformed_body carries the shared guide", () => {
  let message = "";
  try {
    deserializeByContentType("application/json", new TextEncoder().encode("not json"));
    assert.fail("decoding a malformed json body must throw");
  } catch (err) {
    message = String((err as Error).message).toLowerCase();
  }
  for (const keyword of guides.decode_malformed_body.must_contain) {
    assert.ok(
      message.includes(keyword.toLowerCase()),
      `decode_malformed_body guide is missing ${keyword}: ${message}`,
    );
  }
});
